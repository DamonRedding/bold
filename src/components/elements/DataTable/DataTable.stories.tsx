import { action } from '@storybook/addon-actions'
import { storiesOf } from '@storybook/react'
import * as React from 'react'

import { withPropTypes } from '../../../stories-addons/withPropTypes'
import { withTheme } from '../../../stories-addons/withTheme'
import { TableColumn } from '../Table/Table'
import { Text } from '../textual/Text/Text'

import { DataTable, DataTableColumn } from './DataTable'

interface Row {
    id: number
    name: string
    cpf: string
    cns: string
}

const rows: Row[] = [
    { id: 1, name: 'MARIA MACHADO DE JESUS', cns: '000.000', cpf: '000.111.222-33' },
    { id: 2, name: 'MARIA MACHADO DE JESUS', cns: '000.000', cpf: '000.111.222-33' },
    { id: 3, name: 'MARIA MACHADO DE JESUS', cns: '000.000', cpf: '000.111.222-33' },
]

storiesOf('Components', module)
    // .addDecorator(withKnobs)
    .addDecorator(withPropTypes())
    .addDecorator(withTheme())
    .add('DataTable', () => (
        // tslint:disable jsx-no-lambda
        <DataTable sort={['cns,DESC', 'nome']} onSort={action('sort')} rows={rows}>
            <TableColumn
                title='ID'
                render={(row: Row) => (
                    <Text>{row.id}</Text>
                )}
            />
            <DataTableColumn
                title='Profissional'
                name='nome'
                render={(row: Row) => (
                    <Text>{row.name}</Text>
                )}
            />
            <DataTableColumn
                title='CNS'
                name='cns'
                render={(row: Row) => (
                    <Text>{row.cns}</Text>
                )}
            />
            <DataTableColumn
                title='CPF'
                name='cpf'
                render={(row: Row) => (
                    <Text>{row.cpf}</Text>
                )}
            />
            <DataTableColumn
                name='buttons'
                render={(row: Row) => (
                    <div>
                        <a href=''>a</a>
                        {' '}|{' '}
                        <a href=''>b</a>
                    </div>
                )}
            />
        </DataTable>
    ))
