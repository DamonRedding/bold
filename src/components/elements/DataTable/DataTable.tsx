import * as React from 'react'

import { withStyles, WithStylesProps } from '../../../styles'
import { Table, TableColumn, TableColumnProps, TableProps } from '../Table/Table'

export interface DataTableProps<T> extends TableProps<T> {
    sort: string[]
    onSort(sort: string[]): any
}

@withStyles
export class DataTable<T> extends React.PureComponent<DataTableProps<T>> {

    render() {
        const { children, sort, ...tableProps } = this.props
        return (
            <Table {...tableProps}>
                {React.Children.map(children, (child: any, idx) => {
                    if (child && child.type === DataTableColumn) {
                        return (
                            <TableColumn
                                {...child.props}
                                title={this.renderColumnTitle(child.props, sort)}
                            />
                        )
                    } else {
                        return child
                    }
                })}
            </Table>
        )
    }

    private renderColumnTitle = (colProps: DataTableColumnProps<T>, sort: string[]) => {
        const handleClick = () => this.handleTitleClick(colProps)

        if (colProps.title) {
            const dir = this.colOrder(colProps.name, sort)
            const styles = {
                sort: {
                    marginLeft: '0.25rem',
                },
            }

            return (
                <a onClick={handleClick}>
                    {colProps.title || colProps.name}
                    {dir && <Sort styles={styles.sort} dir={dir} />}
                </a>
            )
        } else {
            return null
        }
    }

    private colOrder = (colName: string, currentSort: string[]) => {
        for (const s of currentSort) {
            const parts = s.split(',')
            if (parts[0] === colName) {
                return parts[1] as any || 'ASC'
            }
        }

        return null
    }

    private handleTitleClick = (colProps: DataTableColumnProps<T>) => {
        // TODO: permitir ordenamento múltiplo (segurando shift ao ordenar?)
        this.props.onSort([colProps.name])
    }

}

export interface DataTableColumnProps<T> extends TableColumnProps<T> {
    name: string
}

export class DataTableColumn<T> extends React.Component<DataTableColumnProps<T>> {
    render() {
        return null
    }
}

interface SortProps extends WithStylesProps {
    dir: 'ASC' | 'DESC'
}

@withStyles
class Sort extends React.Component<SortProps> {
    render() {
        const { css, dir } = this.props
        if (dir === 'ASC') {
            return <span className={css()}>▾</span>
        } else if (dir === 'DESC') {
            return <span className={css()}>▴</span>
        } else {
            return null
        }
    }
}
