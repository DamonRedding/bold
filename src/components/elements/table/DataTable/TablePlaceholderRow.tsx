import * as React from 'react'

import { withStyles, WithStylesProps } from '../../../../styles'
import { TableCell, TableRow } from '../Table'

export interface TablePlaceholderRowProps extends WithStylesProps {
    colSpan: number
    message?: string
}

@withStyles
export class TablePlaceholderRow extends React.PureComponent<TablePlaceholderRowProps> {
    static defaultProps: Partial<TablePlaceholderRowProps> = {
        message: 'Nenhum registro encontrado',
    }

    render() {
        const { theme, colSpan, message } = this.props
        const styles = {
            cell: {
                color: theme.color.gray40,
                fontStyle: 'italic',
                textAlign: 'center',
            },
        }
        return (
            <TableRow>
                <TableCell colSpan={colSpan} styles={styles.cell}>
                    {message}
                </TableCell>
            </TableRow>
        )
    }
}
