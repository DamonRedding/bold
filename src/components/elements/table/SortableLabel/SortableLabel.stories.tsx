import { action } from '@storybook/addon-actions'
import { select, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'

import { withPropTypes, withTheme } from '../../../../stories-addons'

import { SortableLabel, SortDirection } from './SortableLabel'

const dirOptions: { [key in SortDirection]: SortDirection } = {
    '': '',
    'ASC': 'ASC',
    'DESC': 'DESC',
}

storiesOf('Components/Table', module)
    .addDecorator(withPropTypes())
    .addDecorator(withTheme())
    .addDecorator(withKnobs)
    .add('SortableLabel', () => (
        <SortableLabel
            dir={select('dir', dirOptions, '')}
            onChange={action('onChange')}
        >
            Property
        </SortableLabel>
    ))
