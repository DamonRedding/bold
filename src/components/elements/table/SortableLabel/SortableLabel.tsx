import * as React from 'react'

import { withStyles, WithStylesProps } from '../../../../styles'
import { Icon } from '../../Icon/Icon'

export type SortDirection = 'ASC' | 'DESC' | ''

export interface SortableLabelProps extends WithStylesProps {
    direction: SortDirection
    onChange(sortDirection: SortDirection, shiftKey?: boolean): any
}

@withStyles
export class SortableLabel extends React.Component<SortableLabelProps> {
    render() {
        const { css, direction } = this.props
        const styles = {
            wrapper: {
                display: 'inline-flex',
                alignItems: 'center',
                cursor: 'pointer',
            },
        }

        const icon = (direction === 'ASC' && 'angleDown')
            || (direction === 'DESC' && 'angleUp')
            || 'sort'

        return (
            <span className={css(styles.wrapper)} onClick={this.handleClick}>
                {this.props.children}
                <Icon
                    style={{ marginLeft: '0.25rem' }}
                    color={icon === 'sort' ? 'disabled' : 'primary'}
                    size={1}
                    icon={icon}
                />
            </span>
        )
    }

    private handleClick = (event: React.MouseEvent<any>) => {
        this.props.onChange(toggleDirection(this.props.direction), event.shiftKey)
    }
}

export const toggleDirection = (dir: SortDirection): 'ASC' | 'DESC' => {
    return (dir === 'ASC' && 'DESC')
        || (dir === 'DESC' && 'ASC')
        || 'ASC'
}
