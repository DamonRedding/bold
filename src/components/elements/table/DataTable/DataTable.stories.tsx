import { action } from '@storybook/addon-actions'
import { storiesOf } from '@storybook/react'
import * as React from 'react'

import { Page } from '../../../../store/requester'
import { withPropTypes, withTheme } from '../../../../stories-addons'
import { Text } from '../../textual/Text/Text'
import { TableColumn } from '../Table/Table'

import { DataTable, DataTableColumn } from './DataTable'

interface Row {
    id: number
    name: string
    cpf: string
    cns: string
}

const rows: Row[] = [
    { id: 1, name: 'MARIA MACHADO DE JESUS', cns: '000.000', cpf: '000.111.222-33' },
    { id: 2, name: 'MARIA MACHADO DE JESUS', cns: '000.000', cpf: '000.111.222-33' },
    { id: 3, name: 'MARIA MACHADO DE JESUS', cns: '000.000', cpf: '000.111.222-33' },
]

const page: Page<Row> = {
    content: rows,
    first: true,
    last: true,
    number: 0,
    numberOfElements: 10,
    totalElements: rows.length,
    size: 10,
    sort: [{ property: 'cns', direction: 'DESC' }, { property: 'nome', direction: 'ASC' }],
    totalPages: 1,
}

storiesOf('Components', module)
    // .addDecorator(withKnobs)
    .addDecorator(withPropTypes())
    .addDecorator(withTheme())
    .add('DataTable', () => (
        // tslint:disable jsx-no-lambda
        <DataTable
            page={page}
            onSortChange={action('sort')}
            onPageChange={action('page-change')}
            onSizeChange={action('size-change')}
        >
            <TableColumn
                title='ID'
                render={(row: Row) => (
                    <Text>{row.id}</Text>
                )}
            />
            <DataTableColumn
                title='Profissional'
                name='nome'
                render={(row: Row) => (
                    <Text>{row.name}</Text>
                )}
            />
            <DataTableColumn
                title='CNS'
                name='cns'
                render={(row: Row) => (
                    <Text>{row.cns}</Text>
                )}
            />
            <DataTableColumn
                title='CPF'
                name='cpf'
                render={(row: Row) => (
                    <Text>{row.cpf}</Text>
                )}
            />
            <DataTableColumn
                name='buttons'
                render={(row: Row) => (
                    <div>
                        <a href=''>a</a>
                        {' '}|{' '}
                        <a href=''>b</a>
                    </div>
                )}
            />
        </DataTable>
    ))
