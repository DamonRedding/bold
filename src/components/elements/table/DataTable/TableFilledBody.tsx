import * as React from 'react'

import { TableBody, TableCell, TableRow } from '../Table'

import { DataTableProps } from './DataTable'
import { TableLoadingRow } from './TableLoadingRow'
import { TablePlaceholderRow } from './TablePlaceholderRow'

export interface TableFilledBodyProps extends Pick<DataTableProps, 'columns' | 'rows' | 'loading' | 'onRowClick'> {

}

export class TableFilledBody extends React.PureComponent<TableFilledBodyProps> {
    render() {
        const { columns, rows, loading, onRowClick } = this.props

        return (
            <TableBody>
                {loading &&
                    <TableLoadingRow colSpan={columns.length} />}

                {!loading && this.isEmpty() &&
                    <TablePlaceholderRow colSpan={columns.length} />}

                {rows.map((row, idx) => (
                    <TableRow
                        key={idx}
                        onClick={onRowClick && this.handleClick(row)}
                    >
                        {columns.map((col, colIdx) => (
                            <TableCell key={colIdx} style={col.styles}>
                                {col.render(row)}
                            </TableCell>
                        ))}
                    </TableRow>
                ))}
            </TableBody>
        )
    }

    private handleClick = (row) => (e) => {
        this.props.onRowClick(row)
    }

    private isEmpty = () => !this.props.rows || this.props.rows.length === 0
}
