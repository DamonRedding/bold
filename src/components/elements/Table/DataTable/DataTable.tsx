import { Interpolation } from 'emotion'
import React from 'react'

import { SortDirection } from '../SortableLabel/SortableLabel'
import { Table, TableHead, TableHeader, TableHeaderProps, TableProps, TableRow } from '../Table'

import { TableFilledBody } from './TableFilledBody'

export interface TableColumnConfig<T = any> {
    name: string
    header?: React.ReactNode
    sortable?: boolean
    style?: Interpolation
    render(row: T): React.ReactNode
}

export interface DataTableProps<T = any> extends TableProps {
    rows: T[]
    columns: Array<TableColumnConfig<T>>
    loading?: boolean
    sort?: string[]
    onSortChange?(sort: string[]): void
    render?(renderProps: DataTableRenderProps): React.ReactNode
    onRowClick?(row: T): any
}

export interface DataTableRenderProps extends DataTableProps {
    getHeaderProps(column: TableColumnConfig | string): TableHeaderProps
    getColumn(columnName: string): TableColumnConfig
}

export class DataTable<T = any> extends React.PureComponent<DataTableProps<T>> {
    static defaultProps: Partial<DataTableProps<any>> = {
        loading: false,
        sort: null,
        onSortChange: () => null,
        render: (renderProps: DataTableRenderProps) => <DataTableDefault {...renderProps} />,
        onRowClick: null,
        hovered: true,
    }

    render() {
        return this.props.render({
            ...this.props,
            getHeaderProps: this.getHeaderProps,
            getColumn: this.getColumn,
        })
    }

    getColumn = (columnName: string): TableColumnConfig => {
        return this.props.columns.find(col => col.name === columnName)
    }

    getHeaderProps = (column: TableColumnConfig | string): TableHeaderProps & { key, 'data-name' } => {
        const col = typeof column === 'string' ? this.getColumn(column) : column

        if (!col) {
            throw new Error(`Column '${column}' not found.`)
        }

        return {
            key: col.name,
            'data-name': col.name,
            sortable: col.sortable,
            sortDirection: this.getSortDirection(col),
            onSortChange: this.handleSortChange(col),
        }
    }

    handleSortChange = (col: TableColumnConfig) => (sortDirection: SortDirection, shiftKey: boolean) => {
        if (shiftKey) {
            this.props.onSortChange(changeSort(this.props.sort, col.name, sortDirection))
        } else {
            this.props.onSortChange([sortDirection === 'ASC' ? col.name : `-${col.name}`])
        }
    }

    getSortDirection = (col: TableColumnConfig): SortDirection => {
        const sorts = this.props.sort || []
        const name = col.name
        for (const sort of sorts) {
            if (sort === name) {
                return 'ASC'
            } else if (sort === `-${name}`) {
                return 'DESC'
            }
        }
        return null
    }
}

const changeSort = (sort: string[], name: string, dir: SortDirection): string[] => {
    const newSort = dir === 'ASC' ? name : `-${name}`
    let swap = false
    const newArray = sort.map(s => {
        if (s === name || s === `-${name}`) {
            swap = true
            return newSort
        }
        return s
    })
    return swap ? newArray : [...newArray, newSort]
}

export class DataTableDefault extends React.PureComponent<DataTableRenderProps> {
    render() {
        const {
            columns,
            rows,
            loading,
            onSortChange,
            sort,
            getHeaderProps,
            getColumn,
            render,
            onRowClick,
            ...rest
        } = this.props

        return (
            <Table {...rest}>
                <TableHead>
                    <TableRow>
                        {columns.map(col => (
                            <TableHeader {...getHeaderProps(col)}>
                                {col.header}
                            </TableHeader>
                        ))}
                    </TableRow>
                </TableHead>
                <TableFilledBody rows={rows} columns={columns} loading={loading} onRowClick={onRowClick} />
            </Table>
        )
    }
}
