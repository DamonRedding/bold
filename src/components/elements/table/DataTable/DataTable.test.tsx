import { mount, render } from 'enzyme'
import * as React from 'react'

import { Page } from '../../../../store/requester'
import { withTheme } from '../../../../test'
import { SortableLabel } from '../SortableLabel/SortableLabel'

import { DataTable, DataTableColumn } from './DataTable'

interface Row {
    id: number
    name: string
}

const rows = [
    { id: 1, name: 'MARIA MACHADO DE JESUS' },
    { id: 2, name: 'MARIA MACHADO DE JESUS' },
]

const page: Page<Row> = {
    content: rows,
    first: true,
    last: true,
    number: 0,
    size: 10,
    numberOfElements: 3,
    totalElements: rows.length,
    sort: [{ property: 'nome', direction: 'ASC' }, { property: 'id', direction: 'DESC' }],
    totalPages: 1,
}

const sortHandler = jest.fn()
const pageHandler = jest.fn()
const sizeHandler = jest.fn()
const table = withTheme(
    // tslint:disable jsx-no-lambda
    <DataTable page={page} onSortChange={sortHandler} onPageChange={pageHandler} onSizeChange={sizeHandler}>
        <DataTableColumn
            name='id'
            sortable={true}
            title='Column ID'
            render={(row: Row) => (
                <span>{row.id}</span>
            )}
        />
        <DataTableColumn
            name='nome'
            sortable={true}
            title='Column name'
            render={(row: Row) => (
                <span>{row.name}</span>
            )}
        />
        <DataTableColumn
            name='actions'
            title='Column actions'
            render={(row: Row) => (
                <span />
            )}
        />
    </DataTable>
)

it('should render correctyle', () => {
    expect(render(table)).toMatchSnapshot()
})

it('should call onSortChange with right parameters when clicked over column title', () => {
    const wrapper = mount(table)
    wrapper.find('th[data-name="id"]').find(SortableLabel).simulate('click')
    expect(sortHandler).toHaveBeenLastCalledWith(['id,ASC'])

    wrapper.find('th[data-name="nome"]').find(SortableLabel).simulate('click')
    expect(sortHandler).toHaveBeenLastCalledWith(['nome,DESC'])

    expect(wrapper.find('th[data-name="actions"]').find(SortableLabel).exists()).toBeFalsy()
})
