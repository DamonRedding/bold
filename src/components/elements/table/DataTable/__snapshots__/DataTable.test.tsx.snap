// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept the render prop 1`] = `<div />`;

exports[`should render correctly 1`] = `
.emotion-27 {
  width: 100%;
  border-collapse: collapse;
  font-size: 0.75rem;
  border: 1px solid #F2F2F7;
}

.emotion-27 tr:hover td {
  background: #F2F2F7;
}

.emotion-3 {
  text-align: left;
  padding: 0.5rem 1rem;
  border-bottom: 1px solid #F2F2F7;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin-left: 0.25rem;
}

.emotion-0 {
  fill: #056DFF;
  font-size: 1rem;
}

.emotion-8 {
  fill: #DCDCE0;
  font-size: 1rem;
}

.emotion-14 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #F2F2F7;
}

.emotion-16 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #F2F2F7;
  text-align: right;
}

<table
  class="emotion-27"
>
  <thead
    class="emotion-12"
  >
    <tr
      class="emotion-12"
    >
      <th
        class="emotion-3"
        data-name="id"
      >
        <span
          class="emotion-2"
        >
          ID
          <span
            class="emotion-1"
          >
            <svg
              aria-hidden="true"
              class="emotion-0"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="M5.64 9.53l1.41-1.42L12 13.06l4.95-4.95 1.41 1.42L12 15.89 5.64 9.53z"
              />
            </svg>
          </span>
        </span>
      </th>
      <th
        class="emotion-3"
        data-name="name"
      >
        <span
          class="emotion-2"
        >
          Name
          <span
            class="emotion-1"
          >
            <svg
              aria-hidden="true"
              class="emotion-0"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="M18.36 14.47l-1.41 1.42L12 10.94l-4.95 4.95-1.41-1.42L12 8.11l6.36 6.36z"
              />
            </svg>
          </span>
        </span>
      </th>
      <th
        class="emotion-3"
        data-name="age"
      >
        <span
          class="emotion-2"
        >
          Age
          <span
            class="emotion-1"
          >
            <svg
              aria-hidden="true"
              class="emotion-8"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="M18.36 8.47l-1.41 1.42L12 4.94 7.05 9.89 5.64 8.47 12 2.11l6.36 6.36zM5.64 15.53l1.41-1.42L12 19.06l4.95-4.95 1.41 1.42L12 21.89l-6.36-6.36z"
              />
            </svg>
          </span>
        </span>
      </th>
    </tr>
  </thead>
  <tbody
    class="emotion-12"
  >
    <tr
      class="emotion-12"
    >
      <td
        class="emotion-14"
      >
        1
      </td>
      <td
        class="emotion-14"
      >
        MARIA MACHADO DE JESUS
      </td>
      <td
        class="emotion-16"
      >
        42
      </td>
    </tr>
    <tr
      class="emotion-12"
    >
      <td
        class="emotion-14"
      >
        2
      </td>
      <td
        class="emotion-14"
      >
        JOSÃ‰ DA SILVA MOREIRA
      </td>
      <td
        class="emotion-16"
      >
        34
      </td>
    </tr>
    <tr
      class="emotion-12"
    >
      <td
        class="emotion-14"
      >
        3
      </td>
      <td
        class="emotion-14"
      >
        ALICE BARBOSA
      </td>
      <td
        class="emotion-16"
      >
        27
      </td>
    </tr>
  </tbody>
</table>
`;
